src/aspire/orientation/commonline_sync.py

```diff
@@ -1,31 +1,443 @@
+
+        # Recover the rotations. The first two columns of all rotation
+        # matrices are given by unmixing V1 and V2 using A. The third
+        # column is the cross product of the first two.
+        r1 = np.dot(a, v1)
+        r2 = np.dot(a, v2)
+        r3 = np.cross(r1, r2, axis=0)
+
+        rotations = np.empty((n_check, 3, 3))
+        rotations[:, :, 0] = r1.T
+        rotations[:, :, 1] = r2.T
+        rotations[:, :, 2] = r3.T
+        # Make sure that we got rotations by enforcing R to be
+        # a rotation (in case the error is large)
+        u, _, v = np.linalg.svd(rotations)
+        np.einsum('ijk, ikl -> ijl', u, v, out=rotations)
+
+        self.rotations = rotations
+
+    def syncmatrix_vote(self):
         """
-        pass
+        Construct the synchronization matrix using voting method
 
-    def estimate(self):
+        A pre-computed common line matrix is required as input.
         """
-        perform estimation of orientations
+        if self.clmatrix is None:
+            self.build_clmatrix()
+
+        clmatrix = self.clmatrix
+
+        sz = clmatrix.shape
+        ell = self.n_theta
+
+        if sz[0] != sz[1]:
+            raise ValueError('clmatrix must be a square matrix.')
+
+        k = sz[0]
+        s = np.eye(2 * k)
+
+        for i in range(k - 1):
+            stmp = np.zeros((2, 2, k))
+            for j in range(i + 1, k):
+                stmp[:, :, j] = self._syncmatrix_ij_vote(
+                    clmatrix, i, j, np.arange(k), ell)
+
+            for j in range(i + 1, k):
+                r22 = stmp[:, :, j]
+                s[2 * i:2 * (i + 1), 2 * j:2 * (j + 1)] = r22
+                s[2 * j:2 * (j + 1), 2 * i:2 * (i + 1)] = r22.T
+
+        self.syncmatrix = s
+
+    def _syncmatrix_ij_vote(self, clmatrix, i, j, k_list, n_theta):
         """
-        pass
+        Compute the (i,j) rotation block of the synchronization matrix using voting method
 
-    def output(self):
+        Given the common lines matrix `clmatrix` and a list of images specified in klist
+        and the number of common lines n_theta.
+        :param clmatrix: The common lines matrix
+        :param i: The i image
+        :param j: The j image
+        :param k_list: The list of images for the third image for voting algorithm
+        :param n_theta: The number of points in the theta direction (common lines)
+        :return: The (i,j) rotation block of the synchronization matrix
         """
-        Output the 3D orientations
+        tol = 1e-12
+
+        good_k, _, _ = self._vote_ij(clmatrix, n_theta, i, j, k_list)
+
+        rs, good_rotations = self._rotratio_eulerangle_vec(
+            clmatrix, i, j, good_k, n_theta)
+
+        if len(good_rotations) > 0:
+            rk = np.mean(rs, 2)
+            tmp_r = rs[:2, :2]
+            diff = tmp_r - rk[:2, :2, np.newaxis]
+            err = np.linalg.norm(diff) / np.linalg.norm(tmp_r)
+            if err > tol:
+                pass
```

Comment:
This means that images i and j have inconsistent rotations. We choose to pass it as Matlab code.

Reply:
But pass does nothing. Why are we even checking?

Reply:
The original Matlab code tried to print out the information on  inconsistent rotations but was commented out and do nothing. Probably due to the fact that it will print out a lots of inconsistent rotations if the resolution or number of images are not enough. Comments added.

Reply:
Comments are fine, but why have a statement that does nothing? What purpose does it serve?

Reply:
The error calculation was removed in the new version.